### ISSUE TITLE
class-string of a union type

### ISSUE THREAD

**Issue by vasily-kartashov** [vimeo/psalm#1851]
Class-strings and union types don't work too well together https://psalm.dev/r/5a6985dce7

```
class A {};
class B {};

/**
 * @template T as A|B
 * @param class-string<T> $type
 * @return T
 */
function f(string $type) {
	return new $type();
}
```

Not sure if it's a bug or by design.

**Comment by muglug** [comment]
Definitely not by design

**Comment by muglug** [comment]
Actually it _is_ by design - `@param class-string<T> $foo` should only expand to an atomic type when calling `new $foo()`. The workaround (which did not previously work, fixed in ba121a9) is to have two different templates:

```php
class A {}
class B {}

/**
 * @template T1 as A
 * @template T2 as B
 * @param class-string<T1>|class-string<T2> $type
 * @return T1|T2
 */
function f(string $type) {
    return new $type();
}

f(A::class);
f(B::class);
f(Exception::class); // error
```

see https://psalm.dev/r/733532308d
